Reject bad string interpolations

beccadax opened the pull request:
String interpolations with multiple comma-separated expressions or argument labels were being incorrectly accepted. This change makes both of them into errors, with appropriate fix-its.

The comma-separated-expressions error could be a warning if we wanted to preserve source compatibility, but as far as I can tell there are no examples of it in the source compatibility suite. The argument-label one can cause an error/crash in serialization if not diagnosed.

Resolves [SR-7937](https://bugs.swift.org/browse/SR-7937) and [SR-7958](https://bugs.swift.org/browse/SR-7958).

beccadax committed:
Reject bad string interpolations

String interpolations with multiple comma-separate expressions or argument labels were being incorrectly accepted.

xwu commented on test/Parse/interpolation_tuple_errors.swift:
```diff
@@ -0,0 +1,23 @@
+
+// Test with too few interpolated values.
+func mamaBear() {
+  _ = "\()" // expected-error{{expected expression in list of expressions}}
```
Any way that we can also make this error less cryptic as part of this change? It’s essentially nonsense right now, at least to me.

brentdax replied:
It was surprisingly nontrivial, but I've just pushed changes to do it. The new version emits the same "interpolations should be a single expression" error message for empty interpolations, but provides a different fix-it telling the user to delete the empty interpolation.

slavapestov commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -1927,6 +1927,37 @@ parseStringSegments(SmallVectorImpl<Lexer::StringSegment> &Segments,
       if (E.isNonNull()) {
         Exprs.push_back(E.get());
+        
+        if (auto tuple = dyn_cast<TupleExpr>(Exprs.back())) {
```
Is the expression ever not a TupleExpr? Maybe use cast<> here instead?

brentdax replied:
The expression actually *should* be a ParenExpr. I don’t know of a reason it should be a TupleExpr without one of the nested conditions being true, but if it ever was, as far as I know it would be valid.

beccadax committed:
Tweak error name to match message

beccadax committed:
Diagnose empty interpolations more clearly

beccadax committed:
Don’t double-diagnose parse errors

Fixes a test at Parse/recovery.swift:799 which the previous commit broke.

beccadax committed:
Fix incorrect test RUN: line

A previous version of this test used FileCheck instead of -verify, and the run line wasn’t properly corrected to use -verify.

beccadax committed:
Update comment

brentdax commented:
@swift-ci Please test

CodaFi commented:
Fingers crossed

@swift-ci please test source compatibility

CodaFi commented on test/Parse/interpolation_tuple_errors.swift:
```diff
@@ -0,0 +1,23 @@
+
+// Test with an argument label
+func funkyBear() {
+  _ = "\(describing: str)" // expected-error{{interpolations cannot start with a keyword argument}} {{10-22=}}
```
A suggestion: Try another `String` initializer or two.

brentdax replied:
Done.

beccadax committed:
Add more argument label tests

Ensures that we don’t get different results from an initializer that doesn’t exist or doesn’t take a String.

beccadax committed:
Resolve the SR-7958 crasher test

We now diagnose the error and remove the label before it has an opportunity to crash.

brentdax commented:
With the crasher in and fixed, I think this one's ready to go.

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test
